### PR TITLE
Move sliced int buffer functionality to MemoryIndex (#11248)

### PR DESCRIPTION
**Pull request by stefanvodita**
`SliceReader`, `SliceWriter`, and all other methods and fields having to do with slices in `IntBlockPool` are moved to `MemoryIndex.SlicedIntBlockPool`, as that is the only consumer of the slices functionality.

There are a few other changes:
1. `IntBlockPool` is no longer `final`, which allows `SlicedIntBlockPool` to extend it.
2. I noticed a buffer pool test I had [committed previously](https://github.com/apache/lucene/pull/12392) was flawed and it only tested overflows for `ByteBlockPool`. I added the same test and corresponding fix for `IntBlockPool` and corrected the assertions.
3. All the existing `IntBlockPool` tests were testing slices. I moved those to their own test class and added a basic write/read/reset test in their place.

Closes #11248

### CODE REVIEW

**Comment by mikemccand** [comment]
Whoa, thanks for tackling this @stefanvodita -- I'll try to review soon.

File: lucene/core/src/java/org/apache/lucene/util/IntBlockPool.java [hunk 1]
```diff
@@ -156,206 +155,6 @@ public void nextBuffer() {
...
-        // We are advancing to the final slice
-        assert end - nextIndex > 0;
-        limit = end - bufferOffset;
-      } else {
-        // This is not the final slice (subtract 4 for the
-        // forwarding address at the end of this new slice)
-        limit = upto + newSize - 1;
-      }
-    }
+    intOffset = Math.addExact(intOffset, INT_BLOCK_SIZE);
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/IntBlockPool.java:158]: Ooh thank you for upgrading to safe addition.

**Review by mikemccand — APPROVED** [review]
Thank you for the new tests, and fixing prior test bugs!  This looks great to me -- nice to move this complexity out of core.

**Comment by stefanvodita** [comment]
@mikemccand, can you merge this change if you think it’s ready?

**Comment by mikemccand** [comment]
I'll backport to 9.x -- this is an internal API so it's fine to change it.

**Comment by mikemccand** [comment]
OK I merged to 9.x as well: https://github.com/apache/lucene/commit/79b9664224f37204504b26e3ed250af01a530f81

THanks @stefanvodita!

**Pull request merged**